src/work/BasicWork.h

```diff
@@ -0,0 +1,114 @@
+// Copyright 2018 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+#pragma once
+
+#include "main/Application.h"
+
+namespace stellar
+{
+
+/** BasicWork is an implementation of a finite state machine,
+ * that is used for async or long-running tasks that:
+ *  - May need to be broken into steps, so as not to block the main thread
+ *  - May fail and need to retry, after some delay
+ *
+ *  BasicWork manages all state transitions via `crankWork` as well as retries.
+ *  While customers can trigger cranking and check on BasicWork's status,
+ *  it is implementers' responsibility to implement `onRun`, that
+ *  hints the next desired state.
+ */
+class BasicWork : public std::enable_shared_from_this<BasicWork>,
+                  private NonMovableOrCopyable
+{
+  public:
+    static size_t const RETRY_NEVER;
+    static size_t const RETRY_ONCE;
+    static size_t const RETRY_A_FEW;
+    static size_t const RETRY_A_LOT;
+    static size_t const RETRY_FOREVER;
+
+    enum State
+    {
+        WORK_RUNNING,
+        WORK_SUCCESS,
+        WORK_WAITING,
+        WORK_FAILURE_RETRY,
+        WORK_FAILURE_RAISE,
+        WORK_DESTRUCTING
+    };
+
+    BasicWork(Application& app, std::function<void()> callback,
```

Comment:
`callback` is not a good name, as it's related to `wakeUp`, maybe `wakeUpCall` gives a bit more of a clue of what it is?